### PR TITLE
[torque-offset-estimator] entity for estimating the offset in the joint torque sensor calibration.

### LINKED ISSUE #13: Joint Torque sensors calibration
We need a calibration procedure to compute the offsets of the torque sensors. This could be carried out by a dedicated entity that takes as input:
- joint encoders
- joint torques
- IMU
- Force/torque sensors (only if we want the calibration to work when the robot is in contact with the environment)

This entity should have a command ```computeOffset``` that triggers the computation of the torque sensor offsets. The offset could be simply computed as the difference between the measurements and the expected joint torques (computed using the dynamic model). In the computation we could assume that the robot velocities and accelerations are null. It will be then up to the user to move the robot into a proper configuration before calling ```computeOffset```.

The output signal of the entity should provide offset-free joint torque measurements.

### PR DESCRIPTION
**Pull request by proyan**
ToDo: Test file

Closes #13

### CODE REVIEW

File: src/torque-offset-estimator.cpp [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+      namespace dynamicgraph = ::dynamicgraph;
+      using namespace dynamicgraph;
+      using namespace dynamicgraph::command;
+      using namespace Eigen;
+
+      /// Define EntityClassName here rather than in the header file
+      /// so that it can be used by the macros DEFINE_SIGNAL_**_FUNCTION.
+      typedef TorqueOffsetEstimator EntityClassName;
+      typedef int dummy;
+      std::vector<Eigen::VectorXd,Eigen::aligned_allocator<Eigen::VectorXd> > stdAlignedVector;
```
Comment (andreadelprete) [inline src/torque-offset-estimator.cpp]: Shouldn't there be a typedef here?

File: src/torque-offset-estimator.cpp [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+        ,CONSTRUCT_SIGNAL_INNER(collectSensorData,  dynamicgraph::Vector, ALL_INPUT_SIGNALS)
+        ,CONSTRUCT_SIGNAL_OUT(jointTorquesEstimated,
+                              dynamicgraph::Vector, m_collectSensorDataSINNER)
+        ,sensor_offset_status(PRECOMPUTATION)
+      {
+        Entity::signalRegistration( ALL_INPUT_SIGNALS << ALL_OUTPUT_SIGNALS);
+        addCommand("init", makeCommandVoid2(*this, &TorqueOffsetEstimator::init,
+                                            docCommandVoid2("Initialize the estimator" ,
+                                                            "urdfFilePath",
+                                                            "Definition of the imu in the chest frame")));
```
Comment (andreadelprete) [inline src/torque-offset-estimator.cpp]: This is not clear, I suggest replacing it with something like "Homogeneous transformation from chest frame to IMU frame"

File: src/torque-offset-estimator.cpp [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+
+      void TorqueOffsetEstimator::computeOffset(const int& nIterations, const double& _epsilon)
+      {
+        if (sensor_offset_status == PRECOMPUTATION) {
+          SEND_MSG("Starting offset computation with no. iterations:"+ nIterations, MSG_TYPE_DEBUG);
+          n_iterations = nIterations;
+          sensor_offset_status = INPROGRESS;
+        }
+        else if (sensor_offset_status == INPROGRESS) {
+          SEND_MSG("Collecting input signals. Please keep the graph running", MSG_TYPE_WARNING_STREAM);
```
Comment (andreadelprete) [inline src/torque-offset-estimator.cpp]: You don't wanna use MSG_TYPE_WARNING_STREAM here because it's not inside the control loop but it's inside a command, so it's gonna be called only once. Replace it with MSG_TYPE_WARNING.

File: src/torque-offset-estimator.cpp [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+          SEND_MSG("Problem in computeOffset sensor_offset_status", MSG_TYPE_ERROR);
+          assert(false && "Problem in computeOffset sensor_offset_status");
+        }
+        return;
+      }
+      
+      DEFINE_SIGNAL_INNER_FUNCTION(collectSensorData, dummy)
+      {
+        if (sensor_offset_status == PRECOMPUTATION)  {
+          SEND_MSG("Currently Precomputation. Not collecting signals", MSG_TYPE_DEBUG);
```
Comment (andreadelprete) [inline src/torque-offset-estimator.cpp]: Here you may wanna use a STREAM msg type

File: src/torque-offset-estimator.cpp [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+        }
+        else if (sensor_offset_status == INPROGRESS) {
+          // Check the current iteration status
+          int _i = encSignals.size();
```
Comment (andreadelprete) [inline src/torque-offset-estimator.cpp]: In general it's better to avoid using variable names that start with underscore because that could mess up with the compiler (not sure if that's true but that's what I've heard).

File: src/torque-offset-estimator.cpp [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+          else {
+            SEND_MSG("This should not happen", MSG_TYPE_ERROR);
+            assert(false);
+          }
+        }
+        else if (sensor_offset_status == COMPUTED) {
+          SEND_MSG("Offset computed", MSG_TYPE_DEBUG);
+        } 
+        else {
+          SEND_MSG("This should not happen", MSG_TYPE_ERROR);
```
Comment (andreadelprete) [inline src/torque-offset-estimator.cpp]: I find all these "This should not happen" a bit pointless. Is there a specific reason for doing that?

File: src/torque-offset-estimator.cpp [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+            SEND_MSG("This should not happen", MSG_TYPE_ERROR);
+            assert(false);
+          }
+          else {
+            SEND_MSG("This should not happen", MSG_TYPE_ERROR);
+            assert(false);
+          }
+        }
+        else if (sensor_offset_status == COMPUTED) {
+          SEND_MSG("Offset computed", MSG_TYPE_DEBUG);
```
Comment (andreadelprete) [inline src/torque-offset-estimator.cpp]: You don't wanna print anything here.

**Review by andreadelprete — CHANGES_REQUESTED** [review]
I think that storing all the data and then doing all the computations at the end is not a good idea. If the number of iterations is sufficiently high that would result in a large computation time at the last iteration. It'd be better to compute the joint torques at each time step, sum them, and at the last iteration you just have to divide by the number of iterations to get the average. Moreover, I think we should stop using N_JOINTS as we are trying to get rid of it to make the code robot independent.

File: src/torque-offset-estimator.cpp (hunk 4 above)
Comment (proyan) [reply src/torque-offset-estimator.cpp]: Since collectSensorData is a inner function, making this STREAM type would completely fill the log output. Perhaps it would be easier if it is called only once for each 'computation state' (precomputation, inprogress or computed)

File: src/torque-offset-estimator.cpp (hunk 6 above)
Comment (proyan) [reply src/torque-offset-estimator.cpp]: I was being a bit careful with the logic in the beginning. I'll remove these

File: src/torque-offset-estimator.cpp (hunk 4 above)
Comment (andreadelprete) [reply src/torque-offset-estimator.cpp]: The STREAM msg types are supposed to be called at each control cycle (but then they are not printed at each control cycle). The other msg types are instead printed every time, so you shouldn't use them inside the signal computations. 

In this specific case I'd say you should just remove this print.

File: src/torque-offset-estimator.cpp [hunk 8]
```diff
@@ -93,10 +93,14 @@ namespace dynamicgraph
...
-          assert(m_model.nv == N_JOINTS+6);
-          jointTorqueOffsets.resize(N_JOINTS);
+          // assert(m_model.nq == N_JOINTS+7);
+          // assert(m_model.nv == N_JOINTS+6);
+
+          jointTorqueOffsets.resize(m_model.nv-6);
           jointTorqueOffsets.setZero();
+
+          ffIndex = 0;
+          torsoIndex = m_model.getJointId("torso_2_joint");
```
Comment (andreadelprete) [inline src/torque-offset-estimator.cpp]: To be general you should take the name of the joint associated to the body containing the IMU ("torso_2_joint") as an input parameter of the init command.

File: src/torque-offset-estimator.cpp [hunk 9]
```diff
@@ -147,48 +143,67 @@ namespace dynamicgraph
...
-            const dynamicgraph::Vector& _enc = m_base6d_encodersSIN(iter);
-            const dynamicgraph::Vector& _acc = m_accelerometerSIN(iter);
-            const dynamicgraph::Vector& _gyr = m_gyroscopeSIN(iter);
-            const dynamicgraph::Vector& _tau = m_jointTorquesSIN(iter);
+            SEND_MSG("Collecting signals for iteration no:" + _i , MSG_TYPE_DEBUG_STREAM);
+
+            const Eigen::VectorXd& sot_enc = m_base6d_encodersSIN(iter);
+            const Eigen::VectorXd& IMU_acc = m_accelerometerSIN(iter);
+            const Eigen::VectorXd& IMU_gyr = m_gyroscopeSIN(iter);
+            const Eigen::VectorXd& _tau = m_jointTorquesSIN(iter);
```
Comment (andreadelprete) [inline src/torque-offset-estimator.cpp]: avoid variable names beginning with underscore

File: src/torque-offset-estimator.cpp [hunk 10]
```diff
@@ -147,48 +143,67 @@ namespace dynamicgraph
...
+
+            //Move the IMU signal to the base frame.
+            //angularAcceleration is zero. Intermediate frame acc and velocities are zero
+            const dynamicgraph::Vector _acc = fMi.rotation()*IMU_acc; //Torso Acceleration
+
+            //Deal with gravity predefined in robot model. Robot Z should be pointing upwards
+            m_model.gravity.linear() = _acc;
+
+            const Eigen::VectorXd& _tau_rnea = se3::rnea(m_model, *m_data, _enc,
+                                                         Eigen::VectorXd::Zero(m_model.njoints),
```
Comment (andreadelprete) [inline src/torque-offset-estimator.cpp]: I guess that njoints is nv-6, am I right? In this case I think that the velocity and acceleration vectors given to RNEA should be of size nv. At the same time I guess that you tested this code before committing it, so it should be working...

File: src/torque-offset-estimator.cpp [hunk 11]
```diff
@@ -147,48 +143,67 @@ namespace dynamicgraph
...
+            const Eigen::VectorXd _current_offset = _tau - _tau_rnea.tail(m_model.njoints);
+            if(_current_offset.array().abs().maxCoeff() >=epsilon) {
+              SEND_MSG("Too high torque offset estimated for iteration"+ _i, MSG_TYPE_ERROR);
+              assert(false);
+            }
+            // encSignals.push_back(_enc);
+            // accSignals.push_back(_acc);
+            // gyrSignals.push_back(_gyr);
+            // tauSignals.push_back(_tau);
+            stdVecJointTorqueOffsets.push_back(_current_offset);
```
Comment (andreadelprete) [inline src/torque-offset-estimator.cpp]: As I said before, there is no need to keep track of all the values computed at each iteration, just sum them all and at the end you just divide by the number of iterations to get the average.

**Review by andreadelprete — CHANGES_REQUESTED** [review]
There are still a few things to fix.

File: include/sot/torque_control/torque-offset-estimator.hh [hunk 12]
```diff
@@ -79,7 +79,6 @@ namespace dynamicgraph {
 
         DECLARE_SIGNAL_IN(base6d_encoders,          dynamicgraph::Vector);
         DECLARE_SIGNAL_IN(accelerometer,            dynamicgraph::Vector);
-        DECLARE_SIGNAL_IN(gyroscope,                dynamicgraph::Vector);
```
Comment (andreadelprete) [inline include/sot/torque_control/torque-offset-estimator.hh]: Why did you remove the gyroscope? I think it may be a good idea to keep it just to check that the robot does not rotate during the calibration

File: src/torque-offset-estimator.cpp [hunk 13]
```diff
@@ -81,22 +80,22 @@ namespace dynamicgraph
...
-                                       const Eigen::Matrix4d& _m_torso_X_imu)
+                                       const Eigen::Matrix4d& m_torso_X_imu_)
       {
         try {
           se3::urdf::buildModel(urdfFile,se3::JointModelFreeFlyer(),m_model);
           // assert(m_model.nq == N_JOINTS+7);
           // assert(m_model.nv == N_JOINTS+6);
 
-          jointTorqueOffsets.resize(m_model.nv-6);
+          jointTorqueOffsets.resize(m_model.nv);
```
Comment (andreadelprete) [inline src/torque-offset-estimator.cpp]: Why did you change the size of the torque offset? I don't see the reason of computing it for the base axes.

File: src/torque-offset-estimator.cpp [hunk 14]
```diff
@@ -126,84 +125,83 @@ namespace dynamicgraph
...
             const Eigen::VectorXd& IMU_acc = m_accelerometerSIN(iter);
-            const Eigen::VectorXd& IMU_gyr = m_gyroscopeSIN(iter);
-            const Eigen::VectorXd& _tau = m_jointTorquesSIN(iter);
+            const Eigen::VectorXd& tau = m_jointTorquesSIN(iter);
 
-            Eigen::VectorXd _enc(m_model.nq);
-            base_sot_to_urdf(sot_enc.head<6>(), _enc.head<7>());
-            _enc.tail(m_model.njoints) = sot_enc.tail(m_model.njoints);
+            Eigen::VectorXd enc(m_model.nq);
+            enc.tail(m_model.nv-6) = sot_enc.tail(m_model.nv-6);
```
Comment (andreadelprete) [inline src/torque-offset-estimator.cpp]: I think here you should use the function joint_sot_to_urdf to convert the joint order from sot to pinocchio convention

File: src/torque-offset-estimator.cpp [hunk 15]
```diff
@@ -126,84 +125,83 @@ namespace dynamicgraph
...
+            base_sot_to_urdf(sot_enc.head<6>(), enc.head<7>());
 
             //Get the transformation from ff(f) to torso (t) to IMU(i) frame:
             // fMi = oMf^-1 * fMt * tMi
-            se3::forwardKinematics(m_model,*m_data,_enc);
+            se3::forwardKinematics(m_model,*m_data,enc);
             se3::SE3 fMi = m_data->oMi[ffIndex].inverse()*m_data->oMi[torsoIndex]*m_torso_X_imu;
```
Comment (andreadelprete) [inline src/torque-offset-estimator.cpp]: instead of computing the inverse transformation you could use the function applyInv, which should be more efficient

File: src/torque-offset-estimator.cpp [hunk 16]
```diff
@@ -126,84 +125,83 @@ namespace dynamicgraph
...
-              SEND_MSG("Too high torque offset estimated for iteration"+ _i, MSG_TYPE_ERROR);
+            m_model.gravity.linear() = acc;
+            
+            //Set fixed for DEBUG
+            //m_model.gravity.linear() = m_model.gravity981;
+
+            const Eigen::VectorXd& tau_rnea = se3::rnea(m_model, *m_data, enc,
+                                                         Eigen::VectorXd::Zero(m_model.nv),
+                                                         Eigen::VectorXd::Zero(m_model.nv));
+            const Eigen::VectorXd current_offset = tau - tau_rnea;
```
Comment (andreadelprete) [inline src/torque-offset-estimator.cpp]: I guess the vector tau has size N, whereas tau_rnea has size N+6, so you cannot subtract them

File: src/torque-offset-estimator.cpp [hunk 17]
```diff
@@ -126,84 +125,83 @@ namespace dynamicgraph
...
+        //        else { // sensor_offset_status == COMPUTED
+        //        } 
         return s; 
       }
 
       DEFINE_SIGNAL_OUT_FUNCTION(jointTorquesEstimated, dynamicgraph::Vector)
       {
         m_collectSensorDataSINNER(iter);
-        if (s.size() != m_model.njoints) s.resize(m_model.njoints);
+        if (s.size() != m_model.nv) s.resize(m_model.nv);
```
Comment (andreadelprete) [inline src/torque-offset-estimator.cpp]: I think the right size for this vector was njoints, why did you change it?

**Review by andreadelprete — CHANGES_REQUESTED** [review]
A few things to change. Did you write a python script to test the torque-offset-estimator entity?

File: src/torque-offset-estimator.cpp (hunk 16 above)
Comment (proyan) [reply src/torque-offset-estimator.cpp]: I wasn't sure that the size of the input sensor is N. Current is assuming that the size of the input sensor is N+6, same as rnea.
I'll change that

File: include/sot/torque_control/torque-offset-estimator.hh (hunk 12 above)
Comment (proyan) [reply include/sot/torque_control/torque-offset-estimator.hh]: I'm not using it anywhere. Do you want to use the gyroscope as a check during run, such that if the check is breached, the program terminates?

File: src/torque-offset-estimator.cpp (hunk 14 above)
Comment (proyan) [reply src/torque-offset-estimator.cpp]: joint order is same in both sot and pinocchio. sot is using pinocchio's urdf parser to make its kinematic tree

File: include/sot/torque_control/torque-offset-estimator.hh (hunk 12 above)
Comment (andreadelprete) [reply include/sot/torque_control/torque-offset-estimator.hh]: Maybe terminating is a bit too much. Just print a message to warn the user that the gyroscope is measuring a significant angular velocity, which might jeopardize the calibration procedure.

File: src/torque-offset-estimator.cpp (hunk 14 above)
Comment (andreadelprete) [reply src/torque-offset-estimator.cpp]: I don't know about Pyrene, but on HRP-2 the order is not the same. Probably @olivier-stasse can confirm whether the joint order is the same between sot and pinocchio for Pyrene.

**Comment by proyan** [comment]
This test is just to confirm the entity. I also don't think we can use this to do the actual experiment :smile_cat:

**Comment by andreadelprete** [comment]
Fair enough, it was just an observation. :)

File: src/torque-offset-estimator.cpp (hunk 14 above)
Comment (olivier-stasse) [reply src/torque-offset-estimator.cpp]: In general the order of the robot motors and the URDF are not the same. We often have to remove some DOFs from the URDF. Thus I would advice to implement the joint_sot_to_urdf.

Comment (proyan) [reply src/torque-offset-estimator.cpp]: I see that joints_sot_to_urdf is in hrp2-common. Is there any other implementation avaliable?

Edit: available for Pyrene?

Comment (olivier-stasse) [reply src/torque-offset-estimator.cpp]: In the refactoring that I am proposing this is provided.
